### PR TITLE
8293630: Simplify TreeMap.get() with Comparator.naturalOrder()

### PR DESCRIPTION
**Pull request by stsypanov**
We can use `Comparator.naturalOrder()` for cases when a `TreeMap` instance is constructed without comparator. This allows to squash two branches in `TreeMap.get()` into one.

P.S. I think the comment of `TreeMap.getEntryUsingComparator()` is outdated. Should we also change it?

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8293630](https://bugs.openjdk.org/browse/JDK-8293630): Simplify TreeMap.get() with Comparator.naturalOrder()

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9901/head:pull/9901` \
`$ git checkout pull/9901`

Update a local copy of the PR: \
`$ git checkout pull/9901` \
`$ git pull https://git.openjdk.org/jdk pull/9901/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9901`

View PR using the GUI difftool: \
`$ git pr show -t 9901`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9901.diff">https://git.openjdk.org/jdk/pull/9901.diff</a>

</details>

### CODE REVIEW

File: src/java.base/share/classes/java/util/TreeMap.java [hunk 1]
```diff
@@ -3322,16 +3314,8 @@ public int characteristics() {
...
-            }
-            else {
-                return (Comparator<Map.Entry<K, V>> & Serializable) (e1, e2) -> {
-                    @SuppressWarnings("unchecked")
-                    Comparable<? super K> k1 = (Comparable<? super K>) e1.getKey();
-                    return k1.compareTo(e2.getKey());
-                };
-            }
+            Comparator<? super K> treeComparator = tree.comparator;
+            return Map.Entry.comparingByKey(treeComparator == null ? natural() : treeComparator);
```
Comment (liach) [inline src/java.base/share/classes/java/util/TreeMap.java:3318]: You can probably have:
```java
return treeComparator == nul ?
        Map.Entry.comparingByKey() :
        Map.Entry.comparingByKey(treeComparator);
```
instead.

Comment (stsypanov) [reply src/java.base/share/classes/java/util/TreeMap.java:3318]: Nope, there'd be a compilation error

**Comment by stsypanov** [comment]
Let's wait

File: src/java.base/share/classes/java/util/TreeMap.java (hunk 1 above)
Comment (ExE-Boss) [reply src/java.base/share/classes/java/util/TreeMap.java:3318]: It'll work fine if an unchecked cast is used:
```suggestion
            @SuppressWarnings("unchecked")
            Comparator<Map.Entry<K, V>> entryComparator = (Comparator) (treeComparator == null ?
                    Map.Entry.comparingByKey()
                    Map.Entry.comparingByKey(treeComparator));
            return entryComparator;
```

This also ensures that when `treeComparator` is `null`, this method keeps returning a constant lambda.

Comment (stsypanov) [reply src/java.base/share/classes/java/util/TreeMap.java]: This one looks fine, accepted.

File: src/java.base/share/classes/java/util/TreeMap.java [hunk 2]
```diff
@@ -3322,16 +3314,12 @@ public int characteristics() {
...
-                return (Comparator<Map.Entry<K, V>> & Serializable) (e1, e2) -> {
-                    @SuppressWarnings("unchecked")
-                    Comparable<? super K> k1 = (Comparable<? super K>) e1.getKey();
-                    return k1.compareTo(e2.getKey());
-                };
-            }
+            Comparator<? super K> treeComparator = tree.comparator;
+            @SuppressWarnings("unchecked")
+            Comparator<Map.Entry<K, V>> entryComparator = treeComparator == null ?
+                    (Comparator) Map.Entry.comparingByKey()
```
Comment (ExE-Boss) [inline src/java.base/share/classes/java/util/TreeMap.java:3320]: Missed a colon:
```suggestion
                    (Comparator) Map.Entry.comparingByKey() :
```

**Comment by amaembo** [comment]
I saw this code many times and always thought that it exists for performance purposes, to avoid extra indirection via likely megamorphic naturalOrder comparator which will slow down the operations on common path. I think such a simplification could be accepted only if accompanied by a properly written benchmark (which actually emulates megamorphic callsite) which shows no performance regression.

File: src/java.base/share/classes/java/util/TreeMap.java [hunk 3]
```diff
@@ -3322,16 +3311,12 @@ public int characteristics() {
         @Override
         public Comparator<Map.Entry<K, V>> getComparator() {
             // Adapt or create a key-based comparator
-            if (tree.comparator != null) {
-                return Map.Entry.comparingByKey(tree.comparator);
-            }
-            else {
-                return (Comparator<Map.Entry<K, V>> & Serializable) (e1, e2) -> {
```
Comment (liach) [inline src/java.base/share/classes/java/util/TreeMap.java:3329]: I think this cast hints compiler to bootstrap lambda meta factory with extra Serializable marker interface. Meanwhile, Comparator.naturalOrder does not implement Serializable.

Comment (stsypanov) [reply src/java.base/share/classes/java/util/TreeMap.java:3329]: From one point this is obviously a change in behaviour, from another the spec of `Spliterator.getComparator()` says nothing about serializability of returned comparator and the actual value is hardly ever going to be serialized.

Comment (ExE-Boss) [reply src/java.base/share/classes/java/util/TreeMap.java:3329]: The only new comparator that is returned here is the one returned by `Map.Entry.comparingByKey()`, which is also `Serializable`.

The only place `Comparator.naturalOrder()` is newly used in this PR are the `final` `package‑private` `getEntry` and `getEntryUsingComparator` methods.

Comment (ExE-Boss) [reply src/java.base/share/classes/java/util/TreeMap.java:3329]: Also, `Comparator.naturalOrder()` is implemented using an `enum`, which is naturally `Serializable`:
https://github.com/openjdk/jdk/blob/0c13d66622a8c2be654bb867aa8c17421d1557ca/src/java.base/share/classes/java/util/Comparators.java#L42-L59

**Comment by stsypanov** [comment]
For the benchmark 
```java
@BenchmarkMode(Mode.AverageTime)
@OutputTimeUnit(TimeUnit.NANOSECONDS)
@Warmup(iterations = 5, time = 1, timeUnit = TimeUnit.SECONDS)
@Measurement(iterations = 15, time = 2, timeUnit = TimeUnit.SECONDS)
@Fork(10)
@State(Scope.Thread)
public class TreeMapGet {

    @Param({"10", "100", "1000", "10000"})
    public int size;

    @Param({"true", "false"})
    public boolean comparator;

    private TreeMap<Integer, Integer> map;
    private Integer[] keys;

    @Setup(Level.Iteration)
    public void setUp() {
        map = comparator ? new TreeMap<>(Comparator.reverseOrder()) : new TreeMap<>();
        keys = IntStream.range(0, size).boxed().toArray(Integer[]::new);
        Collections.reverse(Arrays.asList(keys));
        for (Integer key : keys) {
            map.put(key, key);
        }
    }

    @Benchmark
    public void get(Blackhole bh) {
        var keys = this.keys;
        var map = this.map;
        for (Integer key : keys) {
            bh.consume(map.get(key));
        }
    }
}
```
I got these results:
```
baseline

Benchmark       (comparator)  (size)  Mode  Cnt       Score      Error  Units
TreeMapGet.get          true      10  avgt  150      69.170 ±    0.886  ns/op
TreeMapGet.get          true     100  avgt  150    1369.488 ±   36.312  ns/op
TreeMapGet.get          true    1000  avgt  150   31462.583 ±  261.650  ns/op
TreeMapGet.get          true   10000  avgt  150  490948.456 ± 3395.402  ns/op
TreeMapGet.get         false      10  avgt  150      69.625 ±    0.975  ns/op
TreeMapGet.get         false     100  avgt  150    1076.163 ±   30.976  ns/op
TreeMapGet.get         false    1000  avgt  150   36528.309 ±  140.245  ns/op
TreeMapGet.get         false   10000  avgt  150  422948.634 ±  976.892  ns/op

patch

Benchmark       (comparator)  (size)  Mode  Cnt       Score      Error  Units
TreeMapGet.get          true      10  avgt  150      68.814 ±    1.221  ns/op
TreeMapGet.get          true     100  avgt  150    1280.799 ±   34.319  ns/op
TreeMapGet.get          true    1000  avgt  150   31530.447 ±  329.406  ns/op
TreeMapGet.get          true   10000  avgt  150  435109.171 ± 1807.717  ns/op
TreeMapGet.get         false      10  avgt  150      65.646 ±    0.124  ns/op
TreeMapGet.get         false     100  avgt  150    1083.535 ±   23.545  ns/op
TreeMapGet.get         false    1000  avgt  150   36299.345 ±  494.126  ns/op
TreeMapGet.get         false   10000  avgt  150  444266.085 ± 2315.616  ns/op
```
For benchmark
```java
@BenchmarkMode(Mode.AverageTime)
@OutputTimeUnit(TimeUnit.NANOSECONDS)
@Warmup(iterations = 5, time = 1, timeUnit = TimeUnit.SECONDS)
@Measurement(iterations = 15, time = 2, timeUnit = TimeUnit.SECONDS)
@Fork(10)
@State(Scope.Thread)
public class TreeMapEntrySetStream {

    @Param({"10", "100", "1000", "10000"})
    public int size;

    @Param({"true", "false"})
    public boolean comparator;

    private TreeMap<Integer, Integer> map;

    @Setup(Level.Iteration)
    public void setUp() {
        map = comparator ? new TreeMap<>(Comparator.reverseOrder()) : new TreeMap<>();
        var keys = IntStream.range(0, size).boxed().toArray(Integer[]::new);
        Collections.reverse(Arrays.asList(keys));
        for (Integer key : keys) {
            map.put(key, key);
        }
    }

    @Benchmark
    public void get(Blackhole bh) {
        var map = this.map;
        map.entrySet().stream().forEach(bh::consume);
    }
}
```
results are the following:
```
baseline

Benchmark                  (comparator)  (size)  Mode  Cnt      Score     Error  Units
TreeMapEntrySetStream.get          true      10  avgt  150     32.714 ±   0.117  ns/op
TreeMapEntrySetStream.get          true     100  avgt  150    271.825 ±   0.596  ns/op
TreeMapEntrySetStream.get          true    1000  avgt  150   3267.876 ±  42.650  ns/op
TreeMapEntrySetStream.get          true   10000  avgt  150  43207.384 ± 457.110  ns/op
TreeMapEntrySetStream.get         false      10  avgt  150     30.863 ±   0.071  ns/op
TreeMapEntrySetStream.get         false     100  avgt  150    259.394 ±   0.493  ns/op
TreeMapEntrySetStream.get         false    1000  avgt  150   3362.446 ±  44.108  ns/op
TreeMapEntrySetStream.get         false   10000  avgt  150  42190.313 ± 421.604  ns/op

patch

Benchmark                  (comparator)  (size)  Mode  Cnt      Score     Error  Units
TreeMapEntrySetStream.get          true      10  avgt  150     30.856 ±   0.112  ns/op
TreeMapEntrySetStream.get          true     100  avgt  150    270.719 ±   0.504  ns/op
TreeMapEntrySetStream.get          true    1000  avgt  150   3278.241 ±  41.304  ns/op
TreeMapEntrySetStream.get          true   10000  avgt  150  42549.518 ± 427.607  ns/op
TreeMapEntrySetStream.get         false      10  avgt  150     31.780 ±   0.092  ns/op
TreeMapEntrySetStream.get         false     100  avgt  150    252.551 ±   0.414  ns/op
TreeMapEntrySetStream.get         false    1000  avgt  150   3359.962 ±  44.337  ns/op
TreeMapEntrySetStream.get         false   10000  avgt  150  41613.570 ± 381.721  ns/op
```

**Pull request closed without merging**